SignalServiceKit/src/Messages/OWSMessageManager.m

```diff
@@ -187,23 +187,46 @@ - (void)handleDeliveryReceipt:(OWSSignalServiceProtosEnvelope *)envelope
     OWSAssert(envelope);
     OWSAssert(transaction);
 
-    NSArray<TSOutgoingMessage *> *messages
-        = (NSArray<TSOutgoingMessage *> *)[TSInteraction interactionsWithTimestamp:envelope.timestamp
-                                                                           ofClass:[TSOutgoingMessage class]
-                                                                   withTransaction:transaction];
-    if (messages.count < 1) {
-        // Desktop currently sends delivery receipts for "unpersisted" messages
-        // like group updates, so these errors are expected to a certain extent.
-        DDLogInfo(@"%@ Missing message for delivery receipt: %llu", self.tag, envelope.timestamp);
-    } else {
-        if (messages.count > 1) {
-            DDLogInfo(@"%@ More than one message (%zd) for delivery receipt: %llu",
-                self.tag,
-                messages.count,
-                envelope.timestamp);
-        }
-        for (TSOutgoingMessage *outgoingMessage in messages) {
-            [outgoingMessage updateWithWasDeliveredWithTransaction:transaction];
+    // Old-style delivery notices don't include a "delivery timestamp".
+    [self processDeliveryReceipts:envelope.source
+                   sentTimestamps:@[
+                       @(envelope.timestamp),
+                   ]
+                deliveryTimestamp:nil
+                      transaction:transaction];
+}
+
+// deliveryTimestamp is an optional parameter.
+- (void)processDeliveryReceipts:(NSString *)recipientId
+                 sentTimestamps:(NSArray<NSNumber *> *)sentTimestamps
+              deliveryTimestamp:(NSNumber *_Nullable)deliveryTimestamp
+                    transaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSAssert(recipientId);
+    OWSAssert(sentTimestamps);
+    OWSAssert(transaction);
+
+    for (NSNumber *nsTimestamp in sentTimestamps) {
+        uint64_t timestamp = [nsTimestamp unsignedLongLongValue];
+
+        NSArray<TSOutgoingMessage *> *messages
+            = (NSArray<TSOutgoingMessage *> *)[TSInteraction interactionsWithTimestamp:timestamp
+                                                                               ofClass:[TSOutgoingMessage class]
+                                                                       withTransaction:transaction];
+        if (messages.count < 1) {
+            // Desktop currently sends delivery receipts for "unpersisted" messages
+            // like group updates, so these errors are expected to a certain extent.
+            DDLogInfo(@"%@ Missing message for delivery receipt: %llu", self.tag, timestamp);
```

Comment:
I don't think this is related to desktop since it's the server sending delivery receipts at this point. (and it's the correct behavior IMO)

But this will also discard "early" delivery receipts for a message sent from our other device received before we've processed the sync'd transcript.

Feel free to just make a new card if you don't want to address in this PR.

Reply:
Made a card, added a TODO.